vms/platformvm/txs/executor/proposal_tx_preference.go

```diff
@@ -0,0 +1,188 @@
+// See the file LICENSE for licensing terms.
+
+package executor
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/snow/uptime"
+	"github.com/ava-labs/avalanchego/utils/constants"
+	"github.com/ava-labs/avalanchego/vms/platformvm/reward"
+	"github.com/ava-labs/avalanchego/vms/platformvm/state"
+	"github.com/ava-labs/avalanchego/vms/platformvm/txs"
+)
+
+var (
+	_ txs.Visitor = (*ProposalTxPreference)(nil)
+
+	ErrMissingStakerTx                      = errors.New("failed to get staker transaction")
+	ErrUnexpectedStakerTransactionType      = errors.New("unexpected staker transaction type")
+	ErrStakerWithoutPrimaryNetworkValidator = errors.New("failed to get primary network validator of staker")
+	ErrMissingSubnetTransformation          = errors.New("failed to get subnet transformation")
+	ErrCalculatingUptime                    = errors.New("failed calculating uptime")
+)
+
+type ProposalTxPreference struct {
+	// inputs, to be filled before visitor methods are called
+	PrimaryUptimePercentage float64
+	Uptimes                 uptime.Calculator
+	State                   state.Chain
+	Tx                      *txs.Tx
+
+	// outputs populated by this struct's methods:
+	//
+	// [PrefersCommit] is true iff this node initially prefers to commit the
+	// block containing this transaction.
+	PrefersCommit bool
+}
+
+func (*ProposalTxPreference) CreateChainTx(*txs.CreateChainTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) CreateSubnetTx(*txs.CreateSubnetTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) ImportTx(*txs.ImportTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) ExportTx(*txs.ExportTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) RemoveSubnetValidatorTx(*txs.RemoveSubnetValidatorTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) TransformSubnetTx(*txs.TransformSubnetTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) AddPermissionlessValidatorTx(*txs.AddPermissionlessValidatorTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) AddPermissionlessDelegatorTx(*txs.AddPermissionlessDelegatorTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) TransferSubnetOwnershipTx(*txs.TransferSubnetOwnershipTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) BaseTx(*txs.BaseTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) AddValidatorTx(*txs.AddValidatorTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) AddSubnetValidatorTx(*txs.AddSubnetValidatorTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) AddDelegatorTx(*txs.AddDelegatorTx) error {
+	return ErrWrongTxType
+}
+
+func (*ProposalTxPreference) AdvanceTimeTx(*txs.AdvanceTimeTx) error {
+	return ErrWrongTxType
+}
+
+func (e *ProposalTxPreference) RewardValidatorTx(tx *txs.RewardValidatorTx) error {
```

Comment:
This function is very brittle. Block verification should imply that this returns `nil`. However, this function works over the `lastAccepted` state rather than the `processing` state that it will be verified against.

Reply:
Is this a change in this PR or something that has always worked this way?

Reply:
Yeah, I was wondering moreso if this was new semantic behavior:
```
works over the lastAccepted state rather than the processing state that it will be verified against
```

Reply:
Previously the `State` that was being accessed during block verification was `OnCommit` rather than `lastAccepted`. The previous code did still use `Uptimes` which depends on the `lastAccepted` state.

Reply:
The major difference here is that we are relying on `Options()` to build the child blocks for the `ProposalBlock` (`BuildBlock` never produces `Commit` or `Abort`). So, if we are in a situation where the `ProposalBlock` is verified but the `Options()` call fails globally, then the chain may halt.